crates/ide_assists/src/handlers/replace_derive_with_manual_impl.rs

```diff
@@ -606,6 +606,177 @@ impl Clone for Foo {
+    bin: usize,
+    bar: usize,
+}
+"#,
+            r#"
+struct Foo {
+    bin: usize,
+    bar: usize,
+}
+
+impl PartialEq for Foo {
+    $0fn eq(&self, other: &Self) -> bool {
+        self.bin == other.bin && self.bar == other.bar
+    }
+}
+"#,
+        )
+    }
+
+    #[test]
+    fn add_custom_impl_partial_eq_tuple_struct() {
+        check_assist(
+            replace_derive_with_manual_impl,
+            r#"
+//- minicore: eq
+#[derive(Partial$0Eq)]
+struct Foo(usize, usize);
+"#,
+            r#"
+struct Foo(usize, usize);
+
+impl PartialEq for Foo {
+    $0fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0 && self.1 == other.1
+    }
+}
+"#,
+        )
+    }
+
+    #[test]
+    fn add_custom_impl_partial_eq_empty_struct() {
+        check_assist(
+            replace_derive_with_manual_impl,
+            r#"
+//- minicore: eq
+#[derive(Partial$0Eq)]
+struct Foo;
+"#,
+            r#"
+struct Foo;
+
+impl PartialEq for Foo {
+    $0fn eq(&self, other: &Self) -> bool {
+        true
+    }
+}
+"#,
+        )
+    }
+
+    #[test]
+    fn add_custom_impl_partial_eq_enum() {
+        check_assist(
+            replace_derive_with_manual_impl,
+            r#"
+//- minicore: eq
+#[derive(Partial$0Eq)]
+enum Foo {
+    Bar,
+    Baz,
+}
+"#,
+            r#"
+enum Foo {
+    Bar,
+    Baz,
+}
+
+impl PartialEq for Foo {
+    $0fn eq(&self, other: &Self) -> bool {
+        core::mem::discriminant(self) == core::mem::discriminant(other)
+    }
+}
+"#,
+        )
+    }
+
+    #[test]
+    fn add_custom_impl_partial_eq_tuple_enum() {
+        check_assist(
+            replace_derive_with_manual_impl,
+            r#"
+//- minicore: eq
+#[derive(Partial$0Eq)]
+enum Foo {
+    Bar(String),
+    Baz,
+}
+"#,
+            r#"
+enum Foo {
+    Bar(String),
+    Baz,
+}
+
+impl PartialEq for Foo {
+    $0fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Bar(l0), Self::Bar(r0)) => l0 == r0,
+            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
+        }
+    }
+}
+"#,
+        )
+    }
+
+    #[test]
+    fn add_custom_impl_partial_eq_record_enum() {
+        check_assist(
+            replace_derive_with_manual_impl,
+            r#"
+//- minicore: eq
+#[derive(Partial$0Eq)]
+enum Foo {
+    Bar {
+        bin: String,
+    },
+    Baz {
+        qux: String,
+        fez: String,
+    },
+    Qux {},
+    Bin,
+}
+"#,
+            r#"
+enum Foo {
+    Bar {
+        bin: String,
+    },
+    Baz {
+        qux: String,
+        fez: String,
+    },
+    Qux {},
+    Bin,
+}
+
+impl PartialEq for Foo {
+    $0fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Bar { bin: l_bin }, Self::Bar { bin: r_bin }) => l_bin == r_bin,
+            (Self::Baz { qux: l_qux, fez: l_fez }, Self::Baz { qux: r_qux, fez: r_fez }) => l_qux == r_qux && l_fez == r_fez,
+            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
```

Comment:
This seems like a footgun waiting to happen. If I changed the definition of `Foo` like

```diff
enum Foo {
    Bar {
        bin: String,
    },
    Baz {
        qux: String,
        fez: String,
    },
    Qux {},
    Bin,
+ New (u64),
}
```

Then `Foo::New(0) == Foo::New(1)`. 

It might be better to not use `core::mem::discriminant` with `_`, as when a new varient is added, it will silently just check the same varient, not the data inside.

Note that this also apply to all enums, as they all could add a varient with data

That also raises an interesting question: What should these macros be optimizing for
1. Best code for the current enum
2. Best code for the current enum, and it's future variations
3. Best code for customizing, as if the user is replacing a derive, they probably dont want to do the default thing.

Currently, it seems the code generated is optimizing for 1, and I'm suggesting moving to 2. I suspect 3 is what would actually be best, although I'm not sure what that would look like

Reply:
I based the codegen mostly on what the compiler currently emits. But you make a good point!

I guess matching all enums could just work well enough, and even though it's more code generated, at least we know if folks change it no issues will occur.

Reply:
Actually I don't think we should do that, and here's why. Assuming we change the match statement to include tuples of all variants, we'd be generating the following code:

```rust
impl PartialEq for Foo {
    $0fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Bar { bin: l_bin }, Self::Bar { bin: r_bin }) => l_bin == r_bin,
            (Self::Baz { qux: l_qux, fez: l_fez }, Self::Baz { qux: r_qux, fez: r_fez }) => l_qux == r_qux && l_fez == r_fez,
            (Self::Qux {}, Self::Qux {}) => true,
            (Self::Bin, Self::Bin) => true,
            _ => false,
        }
    }
}
```

Say we add a `New` variant to the enum: the existing `PartialEq` impl will continue to compile. There's a few options we could take: we could generate. The obvious solution would be to manually match every possible case, which for this rather simple enum would end up being __25 match arms__ (n * n). That might be a bit much, so we could create `(Foo, Foo)` and `(Foo, _)` matches, which ends up being __10 match arms__ (2 * n), which is still a lot.

### Conclusion

I think the "best solution" will be different based on what folks want to do with the code. What we're doing now is emitting what the compiler emits, and then letting folks use that as the base for their customizations. I think considering the options, that's the best we can do, and that's what we should stick to for now.

Reply:
Not to drag this discussion out much longer, I also wonder if (despite it potentially being more code and a more complicated implementation), if users will find it more straightforward if the generated code doesn't use `core::mem::discriminant` but instead explicitly matches on every case where the two discriminates are equal. `core::mem::discriminant` isn't super often used, and I imagine most users will likely tend to explicitly match.